lldb/source/Target/SwiftLanguageRuntimeDynamicTypeResolution.cpp

```diff
@@ -1027,22 +1155,117 @@ SwiftLanguageRuntimeImpl::GetPromiseForTypeNameAndFrame(const char *type_name,
+  if (!node)
+    return;
+
+  using namespace swift::Demangle;
+  switch (node->getKind()) {
+  case Node::Kind::DependentGenericParamType: {
+    if (node->getNumChildren() != 2)
+      return;
+    NodePointer depth_node = node->getChild(0);
+    NodePointer index_node = node->getChild(1);
+    if (!depth_node || !depth_node->hasIndex() || !index_node ||
+        !index_node->hasIndex())
+      return;
+    callback(depth_node->getIndex(), index_node->getIndex());
+    break;
+  }
+  default:
+    // Visit the child nodes.
+    for (unsigned i = 0; i < node->getNumChildren(); ++i)
+      ForEachGenericParameter(node->getChild(i), callback);
+  }
+}
+
+CompilerType SwiftLanguageRuntimeImpl::DoArchetypeBindingForTypeRef(
+    StackFrame &stack_frame, TypeSystemSwiftTypeRef &ts,
+    ConstString mangled_name) {
+  auto *reflection_ctx = GetReflectionContext();
+  if (!reflection_ctx)
+    return ts.GetTypeFromMangledTypename(mangled_name);
+
+  swift::Demangle::Demangler Dem;
+  swift::Demangle::NodePointer canonical =
+      TypeSystemSwiftTypeRef::GetCanonicalDemangleTree(
+          ts.GetModule(), Dem, mangled_name.GetStringRef());
+
+  // Build the list of type substitutions.
+  swift::reflection::GenericArgumentMap substitutions;
+  ForEachGenericParameter(canonical, [&](unsigned depth, unsigned index) {
+    if (substitutions.count({depth, index}))
+      return;
+    StreamString mdvar_name;
+    mdvar_name.Printf(u8"$\u03C4_%d_%d", depth, index);
+
+    llvm::Optional<lldb::addr_t> metadata_location =
+        GetTypeMetadataForTypeNameAndFrame(mdvar_name.GetString(), stack_frame);
+    if (!metadata_location)
+      return;
+
+    const swift::reflection::TypeRef *type_ref =
+        reflection_ctx->readTypeFromMetadata(*metadata_location);
+    substitutions.insert({{depth, index}, type_ref});
+  });
+
+  // Nothing to do if there are no type parameters.
+  if (substitutions.empty())
+    return ts.GetTypeFromMangledTypename(ConstString(mangleNode(canonical)));
+
+  // Build a TypeRef from the demangle tree.
+  const swift::reflection::TypeRef *type_ref =
+      swift::Demangle::decodeMangledType(reflection_ctx->getBuilder(),
+                                         canonical);
+
+  // Apply the substitutions.
+  const swift::reflection::TypeRef *bound_type_ref =
+      type_ref->subst(reflection_ctx->getBuilder(), substitutions);
+  swift::Demangle::NodePointer node = bound_type_ref->getDemangling(Dem);
+  CompilerType bound_type = ts.RemangleAsType(Dem, node);
+
+  // Import the type into the scratch context. Subsequent conversions
+  // to Swift types must be performed in the scratch context, since
+  // the bound type may combine types from different
+  // lldb::Modules. Contrary to the AstContext variant of this
+  // function, we don't want to do this earlier, because the
+  // canonicalization in GetCanonicalDemangleTree() must be performed in
+  // the original context as to resolve type aliases correctly.
+  Status error;
```

Comment:
Since we are just pushing strings around there is no way that this could actually fail with a typeref as input. That said — getting the scratch context could. Fixed!